File: java/quarkus-postgres/src/main/java/com/base14/demo/service/ArticleService.java
package com.base14.demo.service;

import com.base14.demo.dto.ArticleDto.ArticleListResponse;
import com.base14.demo.dto.ArticleDto.ArticleResponse;
import com.base14.demo.dto.ArticleDto.CreateArticleRequest;
import com.base14.demo.dto.ArticleDto.UpdateArticleRequest;
import com.base14.demo.entity.Article;
import com.base14.demo.entity.Favorite;
import com.base14.demo.entity.User;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.StatusCode;
import io.opentelemetry.instrumentation.annotations.WithSpan;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import org.jboss.logging.Logger;
import java.text.Normalizer;
import java.time.Instant;
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@ApplicationScoped
public class ArticleService {

    private static final Logger LOG = Logger.getLogger(ArticleService.class);
    private static final Pattern NON_ALPHANUMERIC = Pattern.compile("[^a-z0-9]+");

    @Inject
    TelemetryService telemetry;

    @WithSpan("article.create")
    @Transactional
    public Article create(CreateArticleRequest request, Long authorId) {
        Span span = Span.current();

        User author = User.findById(authorId);
        if (author == null) {
            span.setStatus(StatusCode.ERROR, "author not found");
            throw new ServiceException("author not found", 404);
        }

        String slug = generateSlug(request.title());
        if (Article.existsBySlug(slug)) {
            slug = slug + "-" + Instant.now().toEpochMilli();
        }

        Article article = new Article();
        article.slug = slug;
        article.title = request.title();
        article.description = request.description();
        article.body = request.body();
        article.author = author;
        article.persist();

        telemetry.incrementArticlesCreated();
        span.setStatus(StatusCode.OK, "article created");
        LOG.infof("Article created: %s", article.slug);

        return article;
    }

    @WithSpan("article.getBySlug")
    public Article getBySlug(String slug, Long userId) {
        Span span = Span.current();

        Article article = Article.findBySlug(slug);
        if (article == null) {
            span.setStatus(StatusCode.ERROR, "article not found");
            throw new ServiceException("article not found", 404);
        }

        if (userId != null) {
            article.favorited = Favorite.exists(userId, article.id);
        }

        span.setStatus(StatusCode.OK, "article retrieved");
        return article;
    }

    @WithSpan("article.list")
    public ArticleListResponse list(int limit, int offset, Long userId) {
        Span span = Span.current();

        List<Article> articles = Article.listPaginated(limit, offset);
        long totalCount = Article.count();

        if (userId != null) {
            List<Long> favoriteIds = Favorite.findArticleIdsByUser(userId);
            Set<Long> favoriteSet = favoriteIds.stream().collect(Collectors.toSet());
            articles.forEach(a -> a.favorited = favoriteSet.contains(a.id));
        }

        List<ArticleResponse> responses = articles.stream()
                .map(ArticleResponse::from)
                .toList();

        span.setStatus(StatusCode.OK, "articles listed");
        return new ArticleListResponse(responses, (int) totalCount);
    }

    @WithSpan("article.update")
    @Transactional
    public Article update(String slug, UpdateArticleRequest request, Long userId) {
        Span span = Span.current();

        Article article = Article.findBySlug(slug);
        if (article == null) {
            span.setStatus(StatusCode.ERROR, "article not found");
            throw new ServiceException("article not found", 404);
        }

        if (!article.author.id.equals(userId)) {
            span.setStatus(StatusCode.ERROR, "not authorized");
            throw new ServiceException("not authorized to update this article", 403);
        }

        if (request.title() != null) {
            article.title = request.title();
            article.slug = generateSlug(request.title());
        }
        if (request.description() != null) {
            article.description = request.description();
        }
        if (request.body() != null) {
            article.body = request.body();
        }

        span.setStatus(StatusCode.OK, "article updated");
        LOG.infof("Article updated: %s", article.slug);

        return article;
    }

    @WithSpan("article.delete")
    @Transactional
    public void delete(String slug, Long userId) {
        Span span = Span.current();

        Article article = Article.findBySlug(slug);
        if (article == null) {
            span.setStatus(StatusCode.ERROR, "article not found");
            throw new ServiceException("article not found", 404);
        }

        if (!article.author.id.equals(userId)) {
            span.setStatus(StatusCode.ERROR, "not authorized");
            throw new ServiceException("not authorized to delete this article", 403);
        }

        Favorite.delete("article.id", article.id);
        article.delete();

        telemetry.incrementArticlesDeleted();
        span.setStatus(StatusCode.OK, "article deleted");
        LOG.infof("Article deleted: %s", slug);
    }

    @WithSpan("article.favorite")
    @Transactional
    public Article favorite(String slug, Long userId) {
        Span span = Span.current();

        Article article = Article.findBySlug(slug);
        if (article == null) {
            span.setStatus(StatusCode.ERROR, "article not found");
            throw new ServiceException("article not found", 404);
        }

        if (Favorite.exists(userId, article.id)) {
            span.setStatus(StatusCode.ERROR, "already favorited");
            throw new ServiceException("article already favorited", 409);
        }

        User user = User.findById(userId);
        Favorite favorite = new Favorite();
        favorite.user = user;
        favorite.article = article;
        favorite.persist();

        Article.incrementFavorites(article.id);
        article.favoritesCount++;
        article.favorited = true;

        telemetry.incrementFavoritesAdded();
        span.setStatus(StatusCode.OK, "article favorited");
        LOG.infof("Article favorited: %s by user %d", slug, userId);

        return article;
    }

    @WithSpan("article.unfavorite")
    @Transactional
    public Article unfavorite(String slug, Long userId) {
        Span span = Span.current();

        Article article = Article.findBySlug(slug);
        if (article == null) {
            span.setStatus(StatusCode.ERROR, "article not found");
            throw new ServiceException("article not found", 404);
        }

        if (!Favorite.exists(userId, article.id)) {
            span.setStatus(StatusCode.ERROR, "not favorited");
            throw new ServiceException("article not favorited", 409);
        }

        Favorite.deleteByUserAndArticle(userId, article.id);
        Article.decrementFavorites(article.id);
        article.favoritesCount = Math.max(0, article.favoritesCount - 1);
        article.favorited = false;

        telemetry.incrementFavoritesRemoved();
        span.setStatus(StatusCode.OK, "article unfavorited");
        LOG.infof("Article unfavorited: %s by user %d", slug, userId);

        return article;
    }

    private String generateSlug(String title) {
        String normalized = Normalizer.normalize(title.toLowerCase(), Normalizer.Form.NFD);
        String slug = NON_ALPHANUMERIC.matcher(normalized).replaceAll("-");
        return slug.replaceAll("^-|-$", "");
    }
}

File: csharp/dotnet-sqlserver/src/Api/Services/ArticleService.cs
using System.Diagnostics;
using System.Text.RegularExpressions;
using Api.Data;
using Api.Data.Entities;
using Api.Models;
using Api.Telemetry;
using Microsoft.EntityFrameworkCore;
using OpenTelemetry;
using OpenTelemetry.Trace;

namespace Api.Services;

public partial class ArticleService(AppDbContext context, JobQueue jobQueue, ILogger<ArticleService> logger)
{
    private static readonly ActivitySource ActivitySource = new("DotnetSqlServer.ArticleService");

    public async Task<ArticleResponse> CreateAsync(int userId, CreateArticleRequest request)
    {
        using var activity = ActivitySource.StartActivity("article.create");
        activity?.SetTag("user.id", userId);

        var slug = GenerateSlug(request.Title);

        var article = new Article
        {
            Slug = slug,
            Title = request.Title,
            Description = request.Description,
            Body = request.Body,
            AuthorId = userId
        };

        context.Articles.Add(article);
        await context.SaveChangesAsync();

        activity?.SetTag("article.id", article.Id);
        activity?.SetTag("article.slug", slug);
        AppMetrics.ArticlesCreated.Add(1);

        logger.LogInformation("Article created: {ArticleId} by user {UserId}", article.Id, userId);

        Baggage.SetBaggage("user.id", userId.ToString());
        Baggage.SetBaggage("article.id", article.Id.ToString());
        await jobQueue.EnqueueAsync("notification", new { ArticleId = article.Id, Type = "article_created" });

        var author = await context.Users.FindAsync(userId);
        return ToArticleResponse(article, author!, false);
    }

    public async Task<ArticleListResponse> GetArticlesAsync(int? userId, int limit = 20, int offset = 0)
    {
        using var activity = ActivitySource.StartActivity("article.list");

        var query = context.Articles
            .Include(a => a.Author)
            .OrderByDescending(a => a.CreatedAt)
            .Skip(offset)
            .Take(limit);

        var articles = await query.ToListAsync();
        var totalCount = await context.Articles.CountAsync();

        var userFavorites = userId.HasValue
            ? await context.Favorites
                .Where(f => f.UserId == userId.Value)
                .Select(f => f.ArticleId)
                .ToHashSetAsync()
            : [];

        var responses = articles.Select(a =>
            ToArticleResponse(a, a.Author, userFavorites.Contains(a.Id)));

        return new ArticleListResponse(responses, totalCount);
    }

    public async Task<ArticleResponse?> GetBySlugAsync(string slug, int? userId)
    {
        using var activity = ActivitySource.StartActivity("article.get");
        activity?.SetTag("article.slug", slug);

        var article = await context.Articles
            .Include(a => a.Author)
            .FirstOrDefaultAsync(a => a.Slug == slug);

        if (article is null) return null;

        var favorited = userId.HasValue &&
            await context.Favorites.AnyAsync(f => f.UserId == userId.Value && f.ArticleId == article.Id);

        return ToArticleResponse(article, article.Author, favorited);
    }

    public async Task<ArticleResponse?> UpdateAsync(string slug, int userId, UpdateArticleRequest request)
    {
        using var activity = ActivitySource.StartActivity("article.update");
        activity?.SetTag("article.slug", slug);
        activity?.SetTag("user.id", userId);

        var article = await context.Articles
            .Include(a => a.Author)
            .FirstOrDefaultAsync(a => a.Slug == slug);

        if (article is null) return null;

        if (article.AuthorId != userId)
        {
            logger.LogWarning("Unauthorized update attempt on article {ArticleId} by user {UserId}", article.Id, userId);
            var ex = new UnauthorizedAccessException("Not authorized to update this article");
            activity?.SetStatus(ActivityStatusCode.Error, ex.Message);
            activity?.AddException(ex);
            throw ex;
        }

        if (request.Title is not null)
        {
            article.Title = request.Title;
            article.Slug = GenerateSlug(request.Title);
        }
        if (request.Description is not null) article.Description = request.Description;
        if (request.Body is not null) article.Body = request.Body;
        article.UpdatedAt = DateTime.UtcNow;

        await context.SaveChangesAsync();

        activity?.SetTag("article.id", article.Id);
        AppMetrics.ArticlesUpdated.Add(1);

        logger.LogInformation("Article updated: {ArticleId} by user {UserId}", article.Id, userId);

        var favorited = await context.Favorites.AnyAsync(f => f.UserId == userId && f.ArticleId == article.Id);
        return ToArticleResponse(article, article.Author, favorited);
    }

    public async Task<bool> DeleteAsync(string slug, int userId)
    {
        using var activity = ActivitySource.StartActivity("article.delete");
        activity?.SetTag("article.slug", slug);
        activity?.SetTag("user.id", userId);

        var article = await context.Articles.FirstOrDefaultAsync(a => a.Slug == slug);

        if (article is null) return false;

        if (article.AuthorId != userId)
        {
            logger.LogWarning("Unauthorized delete attempt on article {ArticleId} by user {UserId}", article.Id, userId);
            var ex = new UnauthorizedAccessException("Not authorized to delete this article");
            activity?.SetStatus(ActivityStatusCode.Error, ex.Message);
            activity?.AddException(ex);
            throw ex;
        }

        var articleId = article.Id;
        context.Articles.Remove(article);
        await context.SaveChangesAsync();

        activity?.SetTag("article.id", articleId);
        AppMetrics.ArticlesDeleted.Add(1);

        logger.LogInformation("Article deleted: {ArticleId} by user {UserId}", articleId, userId);

        return true;
    }

    public async Task<ArticleResponse?> FavoriteAsync(string slug, int userId)
    {
        using var activity = ActivitySource.StartActivity("article.favorite");
        activity?.SetTag("article.slug", slug);
        activity?.SetTag("user.id", userId);

        await using var transaction = await context.Database.BeginTransactionAsync();

        try
        {
            var article = await context.Articles
                .Include(a => a.Author)
                .FirstOrDefaultAsync(a => a.Slug == slug);

            if (article is null) return null;

            var existing = await context.Favorites
                .FirstOrDefaultAsync(f => f.UserId == userId && f.ArticleId == article.Id);

            if (existing is null)
            {
                context.Favorites.Add(new Favorite
                {
                    UserId = userId,
                    ArticleId = article.Id
                });
                await context.SaveChangesAsync();

                await context.Articles
                    .Where(a => a.Id == article.Id)
                    .ExecuteUpdateAsync(s => s.SetProperty(a => a.FavoritesCount, a => a.FavoritesCount + 1));

                article.FavoritesCount++;
                AppMetrics.FavoritesAdded.Add(1);

                logger.LogInformation("Article favorited: {ArticleId} by user {UserId}", article.Id, userId);
            }

            await transaction.CommitAsync();

            activity?.SetTag("article.id", article.Id);
            return ToArticleResponse(article, article.Author, true);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to favorite article {Slug} by user {UserId}", slug, userId);
            activity?.SetStatus(ActivityStatusCode.Error, ex.Message);
            activity?.AddException(ex);
            await transaction.RollbackAsync();
            throw;
        }
    }

    public async Task<ArticleResponse?> UnfavoriteAsync(string slug, int userId)
    {
        using var activity = ActivitySource.StartActivity("article.unfavorite");
        activity?.SetTag("article.slug", slug);
        activity?.SetTag("user.id", userId);

        await using var transaction = await context.Database.BeginTransactionAsync();

        try
        {
            var article = await context.Articles
                .Include(a => a.Author)
                .FirstOrDefaultAsync(a => a.Slug == slug);

            if (article is null) return null;

            var favorite = await context.Favorites
                .FirstOrDefaultAsync(f => f.UserId == userId && f.ArticleId == article.Id);

            if (favorite is not null)
            {
                context.Favorites.Remove(favorite);
                await context.SaveChangesAsync();

                await context.Articles
                    .Where(a => a.Id == article.Id && a.FavoritesCount > 0)
                    .ExecuteUpdateAsync(s => s.SetProperty(a => a.FavoritesCount, a => a.FavoritesCount - 1));

                article.FavoritesCount = Math.Max(0, article.FavoritesCount - 1);
                AppMetrics.FavoritesRemoved.Add(1);

                logger.LogInformation("Article unfavorited: {ArticleId} by user {UserId}", article.Id, userId);
            }

            await transaction.CommitAsync();

            activity?.SetTag("article.id", article.Id);
            return ToArticleResponse(article, article.Author, false);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to unfavorite article {Slug} by user {UserId}", slug, userId);
            activity?.SetStatus(ActivityStatusCode.Error, ex.Message);
            activity?.AddException(ex);
            await transaction.RollbackAsync();
            throw;
        }
    }

    private static string GenerateSlug(string title)
    {
        var slug = SlugRegex().Replace(title.ToLowerInvariant(), "");
        slug = WhitespaceRegex().Replace(slug, "-");
        return $"{slug}-{Guid.NewGuid().ToString()[..8]}";
    }

    private static ArticleResponse ToArticleResponse(Article article, User author, bool favorited) =>
        new(
            article.Id,
            article.Slug,
            article.Title,
            article.Description,
            article.Body,
            article.FavoritesCount,
            favorited,
            article.CreatedAt,
            article.UpdatedAt,
            new AuthorResponse(author.Id, author.Name, author.Bio, author.Image)
        );

    [GeneratedRegex("[^a-z0-9\\s-]")]
    private static partial Regex SlugRegex();

    [GeneratedRegex("\\s+")]
    private static partial Regex WhitespaceRegex();
}
